File: langs/python/src/vmx/messages/tree_structure_changed.py
"""TreeStructureChangedMessage (ADR-0028, chapter 18).

Published on the message hub when a HierarchicalVM subtree changes
structurally (add / remove / reparent of a child node).
"""

from __future__ import annotations

from dataclasses import dataclass
from enum import Enum
from typing import Any


class TreeStructureChange(Enum):
    """Discriminated enum for the structural mutation that occurred."""

    ADDED = "added"
    REMOVED = "removed"
    REPARENTED = "reparented"


@dataclass(frozen=True, slots=True)
class TreeStructureChangedMessage:
    """Immutable message emitted on structural mutations in a HierarchicalVM subtree.

    Parameters
    ----------
    sender:
        The node whose ``children`` collection changed.
    sender_name:
        Human-readable identifier for ``sender``.
    change:
        The kind of structural mutation (ADDED / REMOVED / REPARENTED).
    affected:
        The node that was added, removed, or reparented.
    index:
        Position in ``children`` at which the change occurred.
        ``-1`` when not applicable (e.g. REPARENTED).
    """

    sender: Any
    sender_name: str
    change: TreeStructureChange
    affected: Any
    index: int

    @property
    def sender_object(self) -> object:
        """Return the sender as an untyped object (satisfies Message protocol)."""
        return self.sender

    @classmethod
    def create(
        cls,
        sender: Any,
        sender_name: str,
        change: TreeStructureChange,
        affected: Any,
        index: int,
    ) -> TreeStructureChangedMessage:
        """Factory method — equivalent to direct construction."""
        return cls(
            sender=sender,
            sender_name=sender_name,
            change=change,
            affected=affected,
            index=index,
        )

File: langs/typescript/src/messages/treeStructureChanged.ts
/**
 * TreeStructureChangedMessage — published on the message hub when a
 * HierarchicalVM subtree changes structurally (add / remove / reparent).
 *
 * See spec/18-hierarchical-vm.md §6 and ADR-0028 §3.4.
 */
import type { IMessage } from "./types.js";

/** Discriminated union for the structural mutation that occurred. */
export type TreeStructureChange = "added" | "removed" | "reparented";

/**
 * Message published when a HierarchicalVM's children list mutates.
 *
 * @param TSource - Type of the node whose subtree changed.
 * @param TAffected - Type of the node added, removed, or reparented.
 */
export class TreeStructureChangedMessage<TSource, TAffected>
  implements IMessage
{
  readonly sender: TSource;
  readonly senderName: string;
  readonly change: TreeStructureChange;
  readonly affected: TAffected;
  /** Index in the children list at which the change occurred; -1 when not applicable. */
  readonly index: number;

  constructor(
    sender: TSource,
    senderName: string,
    change: TreeStructureChange,
    affected: TAffected,
    index: number,
  ) {
    this.sender = sender;
    this.senderName = senderName;
    this.change = change;
    this.affected = affected;
    this.index = index;
  }
}
